### DISCUSSION TITLE
Can this be used in with tcp ends too

### DISCUSSION THREAD

**Discussion by Stevemoretz** [wangyu-/udp2raw-tunnel#448] [Q&A]
I really like this tool, right now the two ends of the tunnel are udp which is great but sometimes in case you want to just tunnel through a chain of servers you might need the ends of the tunnels to be tcp too, can this be used for that too? if not what's the best alternative?

**Comment by wangyu-** [comment]
### Option 1:
use [kcptun](https://github.com/xtaci/kcptun). kcptun creates a tunnel with both ends as tcp, and kcptun has a build-in fakeTcp mode similiar to udp2raw.

### Option 2:
use udp2raw+ kcptun:
tcp traffic (in) ---> kcptun-client ---> udp2raw-client --------faketcp on internet--------->udp2taw-server--->kcptun-server--->tcp traffic (out)
(some people reported this has better compatbility on their network)
### Option 3:
use udp2raw + wireguard or simliarly udp2raw + openvpn.

**Reply by Stevemoretz** [reply]
@wangyu- Thank you so much! sorry I forgot to mention only ICMP mode works for me :(
But I think option 2 still should work for me right? 🥳 option 1 and 3 can't though I guess

I have two servers so I think this should be mine:

tcp traffic (in) ---> kcptun-client ---> udp2raw-client --------icmp on internet--------->udp2taw-server1 --->kcptun-on-server1 ---> udp2raw-sever1-client------icmp on internet--------->udp2taw-server2--->kcptun-on-server2--->tcp traffic (out)

**Reply by wangyu-** [reply]
>But I think option 2 still should work for me right? 🥳 option 1 and 3 can't though I guess

option 1 and 3 should work as well.     For all the 3, with udp2raw icmp mode, only icmp is exposed to outside.
